src/main/java/tetris/Tetromino.java

```diff
@@ -198,4 +191,61 @@ public Tetromino rotateClockwise(Tetromino shape) {
     }
 
     //add rotateCounter()
+
+    /**
+     * @return last element/bottom of shape.
+     * Used if the shape reached bottom of grid or other shape at bottom.
+     */
+    public Square getShapeBottom(List<Square> shape) {
+        if (oShape.equals(shape)) {
+            return oShape.get(oShape.size()-1);
+        } else if (iShape.equals(shape)) {
+            return iShape.get(iShape.size() - 1);
+        } else if (jShape.equals(shape)) {
+            return jShape.get(jShape.size() - 1);
+        } else if (lShape.equals(shape)) {
+            return lShape.get(lShape.size() - 1);
+        } else if (sShape.equals(shape)) {
+            return sShape.get(sShape.size() - 1);
+        } else if (zShape.equals(shape)) {
+            return zShape.get(zShape.size() - 1);
+        } else if (tShape.equals(shape)) {
+            return tShape.get(tShape.size() - 1);
+        } else if (pShape.equals(shape)) {
+            return pShape.get(pShape.size() - 1);
+        } else if (uShape.equals(shape)) {
+            return uShape.get(uShape.size() - 1);
+        } else if (miniShape.equals(shape)) {
+            return miniShape.get(miniShape.size() - 1);
+        }
+        //if it's an unhandled shape
+        else {
+            return null;
+        }
+    }
+
+    public void getAllShapes() {
+        allShapes.put(1, oShape);
+        allShapes.put(2, iShape);
+        allShapes.put(3, jShape);
+        allShapes.put(4, lShape);
+        allShapes.put(5, sShape);
+        allShapes.put(6, zShape);
+        allShapes.put(7, tShape);
+        allShapes.put(8, pShape);
+        allShapes.put(9, uShape);
+        allShapes.put(10, miniShape);
+    }
+
+    public List<Square> getRandomShape() {
+        //populate HashMap with all shapes
```

Comment:
We shouldn't repopulate the hashmap every time we get a new random shape.

Reply:
Also when we get a shape it needs to make a new ArrayList because each new shape will have different x and y values even if they happen to be the same shape.